### PR TITLE
Add Microsoft Teams support

### LINKED ISSUE #1: Add support for Microsoft Teams?
Ref:

https://adaptivecards.io/
https://messagecardplayground.azurewebsites.net/
https://docs.microsoft.com/en-us/outlook/actionable-messages/send-via-connectors
https://docs.microsoft.com/en-us/microsoftteams/platform/concepts/connectors/connectors-using

### PR DESCRIPTION
**Pull request by ajeet-smpx**
closes #1

### CODE REVIEW

File: teams.js [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+const startCase = require('lodash.startcase');
+
+/**
+ * @typedef {import('@smpx/notify').TeamsTypes.Section} Section
+ * @typedef {import('@smpx/notify').TeamsTypes.Action} Action
+ * @typedef {import('@smpx/notify').TeamsTypes.MessageCard} MessageCard
+ */
+
+let packageObj;
+function getPackageInfo() {
```
Comment (rohit-gohri) [inline teams.js]: could import this from index

File: teams.js [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+	}
+	catch(err) {
+		packageObj = {name: 'Unknown', version: '0.0.0'};
+	}
+	return packageObj;
+}
+
+/** @type {Console} */
+let logger;
+function getLogger() {
```
Comment (rohit-gohri) [inline teams.js]: same for this

File: teams.js (hunk 1 above)
Comment (rohit-gohri) [reply teams.js]: or create a helpers file

File: teams.js [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+class Teams {
+	/**
+	 * Overwrite this function to skip teams message sending in some cnditions
+	 * and log the message instead. By default skips in test environment
+	 */
+	static logCondition() {
+		return cfg.isTest();
+	}
+
+	constructor({text = '', channel = ''} = {}) {
```
Comment (rohit-gohri) [inline teams.js]: can remove default values

File: teams.js [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+	}
+
+	/** @param {string} channel */
+	channel(channel) {
+		this._channel = channel;
+		return this;
+	}
+
+	/** @param {string} summary */
+	summary(summary) {
```
Comment (rohit-gohri) [inline teams.js:110]: Add summary in Slack class too, which sets the fallback in the first attachement

File: teams.js [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+	}
+
+	/** @param {string} text */
+	text(text) {
+		this._text = text;
+		return this;
+	}
+
+	/** @param {Section | Section[]} sections */
+	section(sections) {
```
Comment (rohit-gohri) [inline teams.js]: this could be renamed to attachment

File: teams.js [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+		if (!Array.isArray(sections)) sections = [sections];
+		this._sections = this._sections.concat(sections);
+		return this;
+	}
+
+	/** @param {Action | Action[]} actions */
+	action(actions) {
```
Comment (rohit-gohri) [inline teams.js:145]: Is this equivalent to button?

File: teams.js [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+		});
+		return this;
+	}
+
+	/**
+	 * Add an error as a section
+	 * @param {Error} err
+	 * @param {{label?: string, title?: string}} [param1={}]
+	 */
+	errorSection(err, {label = '', title = ''} = {}) {
```
Comment (rohit-gohri) [inline teams.js]: Could make this static or remove,  no need to for two fns to do similar things

**Review by rohit-gohri — CHANGES_REQUESTED** [review]
Also make dummy functions for username and icon that return this and do nothing

File: teams.js (hunk 1 above)
Comment (ajeet-smpx) [reply teams.js]: :+1:

File: teams.js (hunk 6 above)
Comment (ajeet-smpx) [reply teams.js:145]: No, there is a separate method for `button` which is an `OpenUriAction`

File: teams.js (hunk 3 above)
Comment (ajeet-smpx) [reply teams.js]: typescript error

Comment (rohit-gohri) [reply teams.js]: Make them optional through jsdoc

File: index.d.ts [hunk 8]
```diff
@@ -124,15 +127,168 @@ declare module '@smpx/notify' {
...
 	}
 
 	export default Slack;
 
 	export {
 		MessageAttachment,
 		AttachmentAction,
+		TeamsTypes,
+		Teams,
 	};
```
Comment (rohit-gohri) [inline index.d.ts:293]: Add export for Slack here too

File: slack.js [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+	 * and log the message instead. By default skips in test environment
+	 */
+	static logCondition() {
+		return cfg.isTest();
+	}
+	/**
+	 * Username to send message with, default: 'slackbot'
+	 */
+	static get username() {
+		return 'slackbot';
```
Comment (rohit-gohri) [inline slack.js]: Add setter for this too

File: teams.js [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+	static set defaultChannel(channel) {
+		this._defaultChannel = channel;
+	}
+
+	/** @returns {string} */
+	static get defaultChannel() {
+		return this._defaultChannel || DEFAULT_CHANNEL;
+	}
+
+	username() { return this }
```
Comment (rohit-gohri) [inline teams.js:99]: add comment that Teams doesn't support these options

**Review by rohit-gohri — APPROVED** [review]

**Pull request merged**